lib/dry/system/config/component_dirs.rb

```diff
@@ -78,69 +80,170 @@ class ComponentDirs
+        #
+        #   @param path [String] the path for the component dir, relative to the configured
+        #     container root
+        #   @yieldparam dir [ComponentDir] the component dir to configure
         #
-        # @param path [String] the path for the component dir, relative to the configured
-        #   container root
+        #   @return [ComponentDir] the added component dir
         #
-        # @yieldparam dir [ComponentDir] the component dir to configure
+        #   @example
+        #     component_dirs.add "lib" do |dir|
+        #       dir.default_namespace = "my_app"
+        #     end
         #
-        # @return [ComponentDir] the added component dir
+        #   @see ComponentDir
+        #   @api public
         #
-        # @example
-        #   component_dirs.add "lib" do |dir|
-        #     dir.default_namespace = "my_app"
-        #   end
+        # @overload add(dir)
+        #   Adds a configured component dir
         #
-        # @see ComponentDir
-        def add(path)
+        #   @param dir [ComponentDir] the configured component dir
+        #
+        #   @return [ComponentDir] the added component dir
+        #
+        #   @example
+        #     dir = Dry::System::ComponentDir.new("lib")
+        #     component_dirs.add dir
+        #
+        #   @see ComponentDir
+        #   @api public
+        def add(path_or_dir)
+          path, dir_to_add = path_and_dir(path_or_dir)
+
           raise ComponentDirAlreadyAddedError, path if dirs.key?(path)
 
-          dirs[path] = ComponentDir.new(path).tap do |dir|
-            yield dir if block_given?
+          dirs[path] = dir_to_add.tap do |dir|
+            # Defaults must be applied after yielding, since the dir is being newly added,
+            # and must have its configuration fully in place before we can know which
+            # defaults to apply
+            yield dir if path_or_dir == path && block_given?
             apply_defaults_to_dir(dir)
           end
         end
 
-        # Returns the added component dirs, with default settings applied
+        # Deletes and returns a previously added component dir
         #
-        # @return [Hash<String, ComponentDir>] the component dirs as a hash, keyed by path
-        def dirs
-          @dirs.each { |_, dir| apply_defaults_to_dir(dir) }
+        # @param path [String] the path for the component dir
+        #
+        # @return [ComponentDir] the removed component dir
+        #
+        # @api public
+        def delete(path)
+          dirs.delete(path)
+        end
+
+        # Returns the paths of the component dirs
+        #
+        # @return [Array<String>] the component dir paths
+        #
+        # @api public
+        def paths
+          dirs.keys
+        end
+
+        # Returns the count of component dirs
+        #
+        # @return [Integer]
+        #
+        # @api public
+        def length
+          dirs.length
         end
```

Comment:
can we have a size alias to make it consistent with `Array`? People typically use `size` anyway.

Reply:
@solnic No probs! Just added these :) It's funny, I don't think I've ever personally used `#size` in preference to `#length`, I wonder how I wound up this way, haha

Reply:
@timriley `size` is shorter and easier to type, I **never** use `length` 🙂